src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.java

```diff
@@ -535,76 +574,83 @@ public void removeDocument(String documentId, String parentDocumentId) throws Fi
 
     @Override
     public void deleteDocument(String documentId) throws FileNotFoundException {
-        long docId = Long.parseLong(documentId);
-        updateCurrentStorageManagerIfNeeded(docId);
+        Log.d(TAG, "deleteDocument(), id=" + documentId);
 
-        OCFile file = currentStorageManager.getFileById(docId);
+        Document document = toDocument(documentId);
 
-        if (file == null) {
-            throw new FileNotFoundException("File " + documentId + " not found!");
+        Context context = getContext();
+        if (context == null) {
+            throw new FileNotFoundException("Context may not be null!");
         }
-        Account account = currentStorageManager.getAccount();
 
-        RemoveFileOperation removeFileOperation = new RemoveFileOperation(file.getRemotePath(), false, account, true,
-                                                                          getContext());
+        recursiveRevokePermission(document);
+
+        Document parentFolder = document.getParent();
 
-        RemoteOperationResult result = removeFileOperation.execute(client, currentStorageManager);
+        RemoteOperationResult result = new RemoveFileOperation(document.getRemotePath(), false,
+                                                               document.getAccount(), true, context)
+            .execute(document.getClient(), document.getStorageManager());
 
         if (!result.isSuccess()) {
             throw new FileNotFoundException("Failed to delete document with documentId " + documentId);
         }
+
+        context.getContentResolver().notifyChange(toNotifyUri(parentFolder), null, false);
     }
 
-    @SuppressLint("LongLogTag")
-    private void updateCurrentStorageManagerIfNeeded(long docId) {
-        if (rootIdToStorageManager == null) {
-            try {
-                queryRoots(FileCursor.DEFAULT_DOCUMENT_PROJECTION);
-            } catch (FileNotFoundException e) {
-                Log.e(TAG, "Failed to query roots");
+    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
+    private void recursiveRevokePermission(Document document) {
+        FileDataStorageManager storageManager = document.getStorageManager();
+        OCFile file = document.getFile();
+        if (file.isFolder()) {
+            for (OCFile child : storageManager.getFolderContent(file, false)) {
+                recursiveRevokePermission(new Document(storageManager, child));
             }
         }
 
-        if (currentStorageManager == null ||
-            rootIdToStorageManager.containsKey(docId) && currentStorageManager != rootIdToStorageManager.get(docId)) {
-            currentStorageManager = rootIdToStorageManager.get(docId);
-        }
+        revokeDocumentPermission(document.getDocumentId());
+    }
+
+    @Override
+    public boolean isChildDocument(String parentDocumentId, String documentId) {
```

Comment:
Doc says: Test if a document is descendant (child, grandchild, etc) from the given parent.

So we need something that checks this up to root folder.

```suggestion
     OCFile parentFile = currentStorageManager.getFileById(Long.parseLong(parentDocumentId));
        OCFile childFile = currentStorageManager.getFileById(Long.parseLong(documentId));

        if (childFile.getParentId() == parentFile.getFileId()) {
            return true;
        }

        OCFile currentParent = currentStorageManager.getFileById(childFile.getParentId());

        while (!ROOT_PATH.equals(currentParent.getRemotePath())) {
            if (parentFile.getRemoteId().equals(currentParent.getRemoteId())) {
                return true;
            }

            currentParent = currentStorageManager.getFileById(currentParent.getParentId());
        }
```

Reply:
I am actually getting this exception now which is probably related:
```
E/DatabaseUtils: Writing exception to parcel
    java.lang.SecurityException: Document 861986647/50 is not a descendant of 861986647/6
        at android.provider.DocumentsProvider.enforceTree(DocumentsProvider.java:226)
        at android.provider.DocumentsProvider.query(DocumentsProvider.java:800)
```

Reply:
> Doc says: Test if a document is descendant (child, grandchild, etc) from the given parent.
> 
> So we need something that checks this up to root folder.

You're totally right on this. Thank you.

Reply:
> I am actually getting this exception now which is probably related:

Could you tell which App do you use, which takes advantage of isChildDocument()?

Reply:
I am using an unreleased Android backup app. It doesn't use isChildDocument directly, but it seems the DocumentsProvider is calling this a lot to ensure you have permissions to operate in the current directory.

I am thinking if it is possible to either improve the current algorithm or cache some child/parent relationships, because there seem to be a lot of requests making things slow.

For example: https://paste.fedoraproject.org/paste/ymN8itkeak0NnTJS6uTTNg

Reply:
Could you please test the overall performance of https://github.com/jmue/nextcloud-android/tree/storageprovider-performance ? This branch adds a commit to reduce contentresolver/sql queries.

Reply:
@jmue looks like this branch doesn't exist?